### PR TITLE
STL_Extension: Add bisect_failures debugging utility

### PR DESCRIPTION
**Pull request by lrineau**
## Summary

This PR adds a new debugging utility `bisect_failures` to the STL_Extension package. This function helps identify minimal failing test cases from complex input data by using a bisection-like approach.

**Small feature: [bisect failures debugging utility](https://cgalwiki.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/bisect_failures_debugging_utility)**

## Features

- **Multiple failure type support**: Handles CGAL exceptions, std::exceptions, and exit codes
- **Automatic granularity adjustment**: Starts with ratio=0.5, automatically halves when no fault is found
- **Debug output**: Saves intermediate states (`current`, `bad`, `error`, `final_bad`)
- **Template-based design**: Works with any copyable/assignable data type

## Testing

Added comprehensive test suite (`test_bisect_failures.cpp`) with 14 tests:
- ✅ Edge cases (empty data, no failures)
- ✅ Single element failures (CGAL exception, std::exception, exit code)
- ✅ Pair combinations (all 2-element permutations)
- ✅ Triple combinations (all 3 failure types together)

All tests pass in both Debug and Release configurations.

## Use Case

This utility is particularly useful for:
- Debugging complex geometric algorithms with large input datasets
- Finding minimal reproducible test cases from failing inputs
- Isolating problematic elements in mesh processing, triangulation, etc.

## Example Usage

```cpp
auto get_size = [](const Mesh& m) { return m.number_of_faces(); };
auto simplify = [](Mesh& m, std::size_t start, std::size_t end) -> bool {
  // Remove faces from start to end
  for(std::size_t i = start; i < end && i < m.number_of_faces(); ++i) {
    m.remove_face(start);
  }
  return true;
};
auto run = [](const Mesh& m) -> int { return algorithm(m); };
auto save = [](const Mesh& m, const std::string& prefix) {
  std::ofstream out(prefix + ".off");
  out << m;
};

int result = bisect_failures(mesh, get_size, simplify, run, save);
```

## Files Changed

- `STL_Extension/include/CGAL/bisect_failures.h` (new) - Main utility header
- `STL_Extension/test/STL_Extension/test_bisect_failures.cpp` (new) - Test suite  
- `STL_Extension/test/STL_Extension/CMakeLists.txt` (modified) - Added test

## Checklist

- [x] Code compiles without errors (Debug and Release)
- [x] All tests pass
- [x] Comprehensive documentation with Doxygen comments
- [x] Example usage in documentation
- [x] Follows CGAL coding style conventions
- [x] Deal with false positive in the testsuite

## Release Management

* Affected package(s): `STL_Extension`
* Feature/Small Feature (if any): [bisect failures debugging utility](https://cgalwiki.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/bisect_failures_debugging_utility)
* Link to compiled documentation (obligatory for small feature) [doc](https://cgal.github.io/9118/v2/STL_Extension/group__PkgSTLExtensionUtilities.html#gad56d708faac9063d3de057394a6d8c52)
* License and copyright ownership: no change

### CODE REVIEW

File: STL_Extension/examples/STL_Extension/bisect_failures_example.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2025 GeometryFactory (France).
```
Comment (afabri) [inline STL_Extension/examples/STL_Extension/bisect_failures_example.cpp]: For examples we do not require a copyright header. In the contrary we want it to be taken as a starting point..

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+// This file is part of CGAL (www.cgal.org).
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: LGPL-3.0-or-later OR LicenseRef-Commercial
+//
+//
+// Author(s)     : Laurent Rineau
+//
+// The documentation has been partially generated using Github Copilot.
```
Comment (afabri) [inline STL_Extension/include/CGAL/bisect_failures.h:13]: Only the documentation?   I have no idea what that means for the copyright. Imagine snippets have been learned from sone GPLed code  and then we apply our dual license.

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+ * to test if the failure persists. It starts with a coarse granularity (ratio=0.5, removing
+ * half the elements) and automatically becomes more fine-grained (dividing ratio by 2) when
+ * no fault is found. When a failure is found, it restarts the bisection with the smaller
+ * failing data, progressively narrowing down to the minimal case.
+ *
+ * \tparam InputData The type of input data to bisect (must be copyable and assignable)
+ * \tparam GetSizeFn Function object type: `std::size_t GetSizeFn(const InputData&)`
+ * \tparam SimplifyFn Function object type: `bool SimplifyFn(InputData&, std::size_t start, std::size_t end)`
+ * \tparam RunFn Function object type: `int RunFn(const InputData&)`
+ * \tparam SaveFn Function object type: `void SaveFn(const InputData&, const std::string&)`
```
Comment (afabri) [inline STL_Extension/include/CGAL/bisect_failures.h]:
```suggestion
 * \tparam SaveFn Function object type: `void SaveFn(const InputData& data , const std::string& filename_prefix)`
```

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+ *
+ * \param data The input data to bisect
+ * \param get_size_fn Function that returns the "size" of the data (e.g., number of elements).
+ * \param simplify_fn Function that simplifies the data by removing elements from [start, end).
```
Comment (afabri) [inline STL_Extension/include/CGAL/bisect_failures.h]:
```suggestion
 * \param simplify_fn Function that simplifies the data by removing elements from `[start, end)`.
```

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+ *                    Should return `true` if simplification succeeded, `false` otherwise.
+ * \param run_fn Function that tests the data. Should return 0 (EXIT_SUCCESS) on success, non-zero on failure.
```
Comment (afabri) [inline STL_Extension/include/CGAL/bisect_failures.h]:
```suggestion
 * \param run_fn Function that tests the data. Should return 0 (`EXIT_SUCCESS`) on success, non-zero on failure.
```

Comment (afabri) [reply STL_Extension/include/CGAL/bisect_failures.h]: further to backtick

File: STL_Extension/test/STL_Extension/test_bisect_failures.cpp [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+  std::cout << "## Test 14: Large sparse dataset (200 elements) with {99}\n";
+  Test_data data_200(200);
+  std::iota(data_200.begin(), data_200.end(), 50);
+  // Successfully bisects from 200 elements down to {99}
+  test(14, data_200, true, {99}, {100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149});
+
+  std::cout << "\n✓ All tests passed!\n";
+  return 0;
+}
```
Comment (afabri) [inline STL_Extension/test/STL_Extension/test_bisect_failures.cpp]: missing end of line here?

File: STL_Extension/doc/STL_Extension/PackageDescription.txt [hunk 7]
```diff
@@ -93,6 +93,7 @@
 - `CGAL::Creator_uniform_d<Arg, Result>`
 
 \cgalCRPSection{Utilities}
+- `CGAL::bisect_failures()`
```
Comment (afabri) [inline STL_Extension/doc/STL_Extension/PackageDescription.txt:96]: Is _STL Extensions_ the appropriate package or should it go to _Profiloing Tools...._    
Your opinion @sloriot ?

**Review by afabri — CHANGES_REQUESTED** [review]

File: STL_Extension/include/CGAL/bisect_failures.h (hunk 2 above)
Comment (lrineau) [reply STL_Extension/include/CGAL/bisect_failures.h:13]: Only the documentation. The code is mine.

File: STL_Extension/doc/STL_Extension/PackageDescription.txt (hunk 7 above)
Comment (lrineau) [reply STL_Extension/doc/STL_Extension/PackageDescription.txt:96]: That is not profiling either. But I agree it is a bit far fetched to present it as an extension of
the STL.

Comment (afabri) [reply STL_Extension/doc/STL_Extension/PackageDescription.txt:96]: The profiling chapter is unfortunately a "Miscellaneous"   It is even that the union find and  hash map would better be in STL Extensions.

File: STL_Extension/test/STL_Extension/test_bisect_failures.cpp [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+// This file is part of CGAL (www.cgal.org).
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: GPL-3.0-or-later OR LicenseRef-Commercial
+//
+// Author(s)     : Laurent Rineau
+
+// Test partially written by Github Copilot
+
```
Comment (afabri) [inline STL_Extension/test/STL_Extension/test_bisect_failures.cpp]:
```suggestion
```

File: STL_Extension/doc/STL_Extension/examples.txt [hunk 9]
```diff
@@ -1,4 +1,5 @@
 /*!
+\example STL_Extension/bisect_failures_example.cpp
```
Comment (afabri) [inline STL_Extension/doc/STL_Extension/examples.txt]: Please drop the suffix  `_example`. There are many without it in this directory.

**Comment by afabri** [comment]
Fails in 50% of today's testsuites  https://cgal.geometryfactory.com/CGAL/testsuite/summary-6.2-Ic-40.html?package=STL_Extension_Examples

**Review by lrineau — COMMENTED** [review]
> Fails in 50% of today's testsuites  https://cgal.geometryfactory.com/CGAL/testsuite/summary-6.2-Ic-40.html?package=STL_Extension_Examples 

I have enforced that test to be used without NDEBUG and CGAL_NDEBUG.

**Comment by afabri** [comment]
Segfault in  [CGAL-6.2-Ic-41](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.2-Ic-41/STL_Extension_Examples/TestReport_lrineau_Debian-stable-Release.gz)

**Comment by sloriot** [comment]
Successfully tested in CGAL-6.2-Ic-53

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+#include <iostream>
+#include <optional>
+#include <string>
+
+namespace CGAL {
+
+/**
+ * \ingroup PkgSTLExtensionUtilities
+ *
+ * \brief Bisects input data by iteratively simplifying it to identify the minimal failing case.
```
Comment (sloriot) [inline STL_Extension/include/CGAL/bisect_failures.h]:
```suggestion
 * \brief bisects input data by iteratively simplifying it to identify the minimal failing case.
```

Comment (sloriot) [inline STL_Extension/include/CGAL/bisect_failures.h]: smaller or minimal? from what I understand it is a smaller one, or a local minimal if you prefer

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+ *
+ * This debugging utility helps identify minimal test cases when complex input data causes failures.
+ * It works by iteratively simplifying the data and testing whether the failure persists,
+ * using a bisection-like approach to narrow down to the smallest failing case.
```
Comment (sloriot) [inline STL_Extension/include/CGAL/bisect_failures.h]: note that you could introduce some randomness, by shuffling the indices first. That way you could get different "minimal" cases

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+ *
+ * \tparam InputData The type of input data to bisect (must be copyable and assignable)
+ * \tparam GetSizeFn Function object type: `std::size_t GetSizeFn(const InputData& data)`
+ * \tparam SimplifyFn Function object type: `bool SimplifyFn(InputData& data, std::size_t start, std::size_t end)`
+ * \tparam RunFn Function object type: `int RunFn(const InputData& data)`
+ * \tparam SaveFn Function object type: `void SaveFn(const InputData& data, const std::string& filename_prefix)`
+ *
+ * \param data The input data to bisect
+ * \param get_size_fn Function that returns the "size" of the data (e.g., number of elements).
+ * \param simplify_fn Function that simplifies the data by removing elements from `[start, end)`.
```
Comment (sloriot) [inline STL_Extension/include/CGAL/bisect_failures.h]:
```suggestion
 * \param simplify_fn Function that simplifies the data by removing elements with indices in `[start, end)`.
```

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+ *                    Should return `true` if simplification succeeded, `false` otherwise.
+ * \param run_fn Function that tests the data. Should return 0 (`EXIT_SUCCESS`) on success, non-zero on failure.
+ *                May also throw exceptions to indicate failure.
+ * \param save_fn Function that saves the data to a file or output. Its second parameter
```
Comment (sloriot) [inline STL_Extension/include/CGAL/bisect_failures.h]: you should indicate that this is the range of possible values, so that it is clear for example that it's up to the user to handle the several "bad" if the intermediate results must be saved (internal integer increased).

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+ * to test if the failure persists. It starts with a coarse granularity (ratio=0.5, removing
+ * half the elements) and automatically becomes more fine-grained (dividing ratio by 2) when
+ * no fault is found. When a failure is found, it restarts the bisection with the smaller
+ * failing data, progressively narrowing down to the minimal case.
+ *
+ * \tparam InputData The type of input data to bisect (must be copyable and assignable)
+ * \tparam GetSizeFn Function object type: `std::size_t GetSizeFn(const InputData& data)`
+ * \tparam SimplifyFn Function object type: `bool SimplifyFn(InputData& data, std::size_t start, std::size_t end)`
+ * \tparam RunFn Function object type: `int RunFn(const InputData& data)`
+ * \tparam SaveFn Function object type: `void SaveFn(const InputData& data, const std::string& filename_prefix)`
```
Comment (sloriot) [inline STL_Extension/include/CGAL/bisect_failures.h]: Maybe another name should be picked. It is not really about saving but a call at the end of an iteration. Similarly, an enum might be more suited rather than a string (more natural for switch-case).

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+ * \param run_fn Function that tests the data. Should return 0 (`EXIT_SUCCESS`) on success, non-zero on failure.
+ *                May also throw exceptions to indicate failure.
+ * \param save_fn Function that saves the data to a file or output. Its second parameter
+ *                (`filename_prefix`) indicates the context (e.g., "bad", "final_bad", "error", "current")
+ *                and can be used to name the output accordingly.
+ *
+ * \return Exit code: 0 (EXIT_SUCCESS) if no failures found, non-zero otherwise
+ *
+ * The algorithm:
+ * 1. Tests the full data first to verify it fails and capture the failure pattern
```
Comment (sloriot) [inline STL_Extension/include/CGAL/bisect_failures.h]:
```suggestion
 * 1. Tests the full data first to verify if it fails and captures the failure pattern
```

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+ * 2. Starts with a ratio of 0.5 (removing 50% of elements) and divides data into "buckets"
+ * 3. For each bucket, creates a simplified version by removing that portion
+ * 4. Tests the simplified version with `run_fn`
+ * 5. If it fails with the same pattern as the original, saves it as "bad" and restarts bisection with this smaller dataset
+ * 6. If it succeeds or fails differently, tries the next bucket
+ * 7. After a complete pass with no matching failures found, reduces the ratio by half (0.5 → 0.25 → 0.125...)
+ * 8. Repeats until no further simplification is possible (minimal failing case found)
+ * 9. Saves the minimal failing case as "final_bad" and returns its exit code
```
Comment (sloriot) [inline STL_Extension/include/CGAL/bisect_failures.h]: Calls `save_fn` ...

File: STL_Extension/include/CGAL/bisect_failures.h (hunk 15 above)
Comment (sloriot) [reply STL_Extension/include/CGAL/bisect_failures.h]: what if it does not fail? directly exit?

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+ *                May also throw exceptions to indicate failure.
+ * \param save_fn Function that saves the data to a file or output. Its second parameter
+ *                (`filename_prefix`) indicates the context (e.g., "bad", "final_bad", "error", "current")
+ *                and can be used to name the output accordingly.
+ *
+ * \return Exit code: 0 (EXIT_SUCCESS) if no failures found, non-zero otherwise
+ *
+ * The algorithm:
+ * 1. Tests the full data first to verify it fails and capture the failure pattern
+ * 2. Starts with a ratio of 0.5 (removing 50% of elements) and divides data into "buckets"
```
Comment (sloriot) [inline STL_Extension/include/CGAL/bisect_failures.h]: The description is a bit too linear, and it is not clear what is the simplification path. For example, I guess the reduction ratio is local. That is when you partition a bucket, you also save its parent and its reduction ratio so that if the bucket "fails", you divide the parent reduction ratio by 2 to generate new son buckets.

File: STL_Extension/include/CGAL/bisect_failures.h (hunk 10 above)
Comment (lrineau) [reply STL_Extension/include/CGAL/bisect_failures.h]: That is actually a local minimum. I wonder how one can document that.

Comment (afabri) [reply STL_Extension/include/CGAL/bisect_failures.h]: Maybe something like "to identify a failing case, from which no further elements can be removed while still failing"

File: STL_Extension/include/CGAL/bisect_failures.h (hunk 14 above)
Comment (lrineau) [reply STL_Extension/include/CGAL/bisect_failures.h]: I actually use the `filename_prefix` as a prefix for the file name to save to.

An enum might work as well. I can do the change.

Comment (sloriot) [reply STL_Extension/include/CGAL/bisect_failures.h]: It goes with my remark that maybe the function shouldn't be called save_fn but something related to the fact that it's called at the end of an iteration

**Comment by sloriot** [comment]
Seems to fail with ICPC. See [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.2-Ic-82/STL_Extension/TestReport_Christo_ICPC-2025.1-Release.gz)

File: STL_Extension/include/CGAL/bisect_failures.h (hunk 14 above)
Comment (afabri) [reply STL_Extension/include/CGAL/bisect_failures.h]: Similar to the BGL visitors ?   Like   `before_something()`      
Also `_fn`   is too short.

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+ * \param get_size_fn Function that returns the "size" of the data (e.g., number of elements).
+ * \param simplify_fn Function that simplifies the data by removing elements with indices in `[start, end)`.
+ *                    Should return `true` if simplification succeeded, `false` otherwise.
+ * \param run_fn Function that tests the data. Should return 0 (`EXIT_SUCCESS`) on success, non-zero on failure.
+ *                May also throw exceptions to indicate failure.
+ * \param save_fn Function that saves the data to a file or output. Its second parameter
+ *                (`filename_prefix`) indicates the context (e.g., "bad", "final_bad", "error", "current")
+ *                and can be used to name the output accordingly.
+ *
+ * \return Exit code: 0 (EXIT_SUCCESS) if no failures found, non-zero otherwise
```
Comment (afabri) [inline STL_Extension/include/CGAL/bisect_failures.h]:
```suggestion
 * \return Exit code: 0 (`EXIT_SUCCESS`) if no failures found, non-zero otherwise
```

File: STL_Extension/include/CGAL/bisect_failures.h (hunk 14 above)
Comment (lrineau) [reply STL_Extension/include/CGAL/bisect_failures.h]: @afabri If `*_fn` is too short, what would you call them?

File: STL_Extension/include/CGAL/bisect_failures.h (hunk 17 above)
Comment (lrineau) [reply STL_Extension/include/CGAL/bisect_failures.h]: I have modified the description of the algorithm. It is still too linear. Please make suggestions.

File: STL_Extension/include/CGAL/bisect_failures.h (hunk 14 above)
Comment (afabri) [reply STL_Extension/include/CGAL/bisect_failures.h]: Usually we write things out.   So probably  `_function`  ? I am not even sure if fn means function or file name

Comment (lrineau) [reply STL_Extension/include/CGAL/bisect_failures.h]: Actually, in the STL, the function object argument name do not contain any hint of them being
functions or function-object, see for example `Pred pred, Proj proj` in the documentation of [`std::ranges::find_if`](https://en.cppreference.com/w/cpp/algorithm/ranges/find).

Maybe `GetSize`, `Simplify`, `Run`,  and `Save` (or another name for the latter) would be better.

**Comment by lrineau** [comment]
> Seems to fail with ICPC. See [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.2-Ic-82/STL_Extension/TestReport_Christo_ICPC-2025.1-Release.gz)

ICPC version is 2025.1 on Christo, whereas the latest version is 2025.3. Maybe we should upgrade.

**Comment by lrineau** [comment]
@afabri @sloriot Please re-review: [`CGAL bisect_failures` v2](https://cgal.github.io/9118/v2/STL_Extension/group__PkgSTLExtensionUtilities.html#gad56d708faac9063d3de057394a6d8c52).

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+ *    - and tests the simplified version by a call `run(data)`.
+ *
+ *    Then:
+ *    - If it fails with the same pattern as the original, calls `notify(data, BAD_DATA)` and restarts bisection
+ *       with this smaller dataset and the same ratio.
+ *    - If it fails differently, calls `notify(data, ERROR_DATA)` and continues with the next bucket.
+ *    - If it succeeds, continue with the next bucket.
+ * 4. After a complete pass with no matching failures found, reduces the ratio by half (0.5 → 0.25 → 0.125...).
+ * 5. Repeats until no further simplification is possible (minimal failing case found).
+ * 6. Calls `notify(data, FINAL_BAD_DATA)` with the minimal failing case and return the result of `run(data)` on it.
```
Comment (sloriot) [inline STL_Extension/include/CGAL/bisect_failures.h]:
```suggestion
 * 6. Calls `notify(min_data, FINAL_BAD_DATA)` with the minimal failing case `min_data` and returns the result of `run(min_data)`.
```

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 20]
```diff
@@ -0,0 +1,288 @@
...
+ *
+ * \warning CGAL::bisect_failures requires the tested code to be compiled with
```
Comment (sloriot) [inline STL_Extension/include/CGAL/bisect_failures.h]:
```suggestion
 * \warning `CGAL::bisect_failures()` requires the tested code to be compiled with
```

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+ * assertions enabled. That means NDEBUG and CGAL_NDEBUG should not be defined. If `run` fails
```
Comment (sloriot) [inline STL_Extension/include/CGAL/bisect_failures.h]:
```suggestion
 * assertions enabled. That means `NDEBUG` and `CGAL_NDEBUG` should not be defined. If `run` fails
```

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 22]
```diff
@@ -0,0 +1,288 @@
...
+ * with a segmentation fault, This function cannot catch it and will also crash.
+ *
+ * Here is an example of how to use `CGAL::bisect_failures`:
```
Comment (sloriot) [inline STL_Extension/include/CGAL/bisect_failures.h]:
```suggestion
 * Here is an example of how to use `CGAL::bisect_failures()`:
```

**Review by sloriot — COMMENTED** [review]
I like this version better! Still some occurrences of minimal and smallest remaining that I'd prefer being replaced with reduced for example.

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 23]
```diff
@@ -0,0 +1,288 @@
...
+ * to test if the failure persists. It starts with a coarse granularity (ratio=0.5, removing
+ * half the elements) and automatically becomes more fine-grained (dividing ratio by 2) when
+ * no fault is found. When a failure is found, it restarts the bisection with the smaller
+ * failing data, progressively narrowing down to the minimal case.
+ *
+ * \tparam InputData The type of input data to bisect (must be copyable and assignable)
+ * \tparam GetSize Function object type, callable with a signature `std::size_t GetSize(const InputData& data)`
+ * \tparam Simplify Function object type, callable with a signature `bool Simplify(InputData& data, std::size_t start, std::size_t end)`
+ * \tparam Run Function object type, callable with a signature `int Run(const InputData& data)`
+ * \tparam Notify Function object type, callable with a signature `void Notify(const InputData& data, Bisection_event event)`
```
Comment (afabri) [inline STL_Extension/include/CGAL/bisect_failures.h]: model of `Callable` ??

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 24]
```diff
@@ -0,0 +1,288 @@
...
+ *
+ * \param data The input data to bisect
+ * \param get_size Function that returns the "size" of the data (e.g., number of elements).
+ * \param simplify Function that simplifies the data by removing elements with indices in `[start, end)`.
+ *                 Should return `true` if simplification succeeded, `false` otherwise.
+ * \param run Function that tests the data. Should return 0 (`EXIT_SUCCESS`) on success, non-zero on failure.
+ *            May also throw exceptions to indicate failure.
+ * \param notify Function that is called with the data at different stages. It can be used to save
+ *                the data to a file or output. Its second parameter `event`, of type `Bisection_event`,
```
Comment (afabri) [inline STL_Extension/include/CGAL/bisect_failures.h:76]: Can we somehow generate a link to the `enum`  @albert-github ?

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 25]
```diff
@@ -0,0 +1,288 @@
...
+  FINAL_BAD_DATA     ///< Minimal failing data that cannot be further simplified
+};
+
+/**
+ * \ingroup PkgSTLExtensionUtilities
+ *
+ * \brief bisects input data by iteratively simplifying it to identify a failing case, from which no
+ *        further elements can be removed while still failing.
+ *
+ * This debugging utility helps identify minimal test cases when complex input data causes failures.
```
Comment (afabri) [inline STL_Extension/include/CGAL/bisect_failures.h]:
```suggestion
 * This debugging utility helps to reduce the size of test cases when complex input data causes failures.
```

File: STL_Extension/include/CGAL/bisect_failures.h (hunk 19 above)
Comment (afabri) [reply STL_Extension/include/CGAL/bisect_failures.h]: Here I find the word "minimal"   ok.

File: STL_Extension/examples/STL_Extension/bisect_failures.cpp [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+    std::string out_filename = prefix + ".off";
+    if(!CGAL::IO::write_polygon_mesh(out_filename, m)) {
+      std::cerr << "Warning: Could not save mesh to " << out_filename << std::endl;
+    } else {
+      std::cout << "Saved mesh with " << m.number_of_faces()
+                << " faces to " << out_filename << std::endl;
+    }
+  };
+
+  // Run bisection to find minimal failing case
```
Comment (afabri) [inline STL_Extension/examples/STL_Extension/bisect_failures.cpp]:
```suggestion
  // Run bisection to find smaller failing case
```

File: STL_Extension/examples/STL_Extension/bisect_failures.cpp [hunk 27]
```diff
@@ -0,0 +1,110 @@
...
+  std::cout << "\n=== Starting bisection to find minimal failing case ===\n" << std::endl;
```
Comment (afabri) [inline STL_Extension/examples/STL_Extension/bisect_failures.cpp]:
```suggestion
  std::cout << "\n=== Starting bisection to find smaller failing case ===\n" << std::endl;
```

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 28]
```diff
@@ -0,0 +1,288 @@
...
+#include <optional>
+#include <string>
+#include <tuple>
+
+namespace CGAL {
+
+/**
+ * \ingroup PkgSTLExtensionUtilities
+ * \brief Enumeration of events during the bisection process
+ * \sa bisect_failures
```
Comment (afabri) [inline STL_Extension/include/CGAL/bisect_failures.h]:
```suggestion
 * \sa bisect_failures()
```

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 29]
```diff
@@ -0,0 +1,288 @@
...
+
+/**
+ * \ingroup PkgSTLExtensionUtilities
+ *
+ * \brief bisects input data by iteratively simplifying it to identify a failing case, from which no
+ *        further elements can be removed while still failing.
+ *
+ * This debugging utility helps identify minimal test cases when complex input data causes failures.
+ * It works by iteratively simplifying the data and testing whether the failure persists,
+ * using a bisection-like approach to narrow down to the smallest failing case.
```
Comment (afabri) [inline STL_Extension/include/CGAL/bisect_failures.h]:
```suggestion
 * using a bisection-like approach to narrow down to a failing case that can not be further reduced.
```

File: STL_Extension/include/CGAL/bisect_failures.h [hunk 30]
```diff
@@ -0,0 +1,288 @@
...
+ *
+ * The algorithm divides the input data into "buckets" and systematically removes each bucket
+ * to test if the failure persists. It starts with a coarse granularity (ratio=0.5, removing
+ * half the elements) and automatically becomes more fine-grained (dividing ratio by 2) when
+ * no fault is found. When a failure is found, it restarts the bisection with the smaller
+ * failing data, progressively narrowing down to the minimal case.
```
Comment (afabri) [inline STL_Extension/include/CGAL/bisect_failures.h]:
```suggestion
 * failing data.
```

**Review by afabri — CHANGES_REQUESTED** [review]
I made suggestions for replacing minimal @lrineau

File: STL_Extension/include/CGAL/bisect_failures.h (hunk 24 above)
Comment (lrineau) [reply STL_Extension/include/CGAL/bisect_failures.h:76]: with Doxygen 1.16, I have the link to `Bisection_event`.

**Review by lrineau — COMMENTED** [review]
I have applied all of your suggestions in the last commit.

File: STL_Extension/include/CGAL/bisect_failures.h (hunk 24 above)
Comment (albert-github) [reply STL_Extension/include/CGAL/bisect_failures.h:76]: Sorry for not responding earlier, been away.

So problem solved when using 1.16.1 (or higher) as I understand it.

Comment (afabri) [reply STL_Extension/include/CGAL/bisect_failures.h:76]: Why does the CI not use this ?

Comment (afabri) [reply STL_Extension/include/CGAL/bisect_failures.h:76]: Not the CI, but our   "build the documentation" command.

Comment (albert-github) [reply STL_Extension/include/CGAL/bisect_failures.h:76]: @afabri would defer from the official release documentation (still using 1.9.6 and is one of the overnight builds as well), would be a nice enhancement possibility for the "build the documentation)_

Comment (lrineau) [reply STL_Extension/include/CGAL/bisect_failures.h:76]: The CI uses the same Doxygen version as the one we use for the official releases. So far, that is Doxygen 1.9.6.

We should bump the Doxygen version for CGAL-6.2 to a more recent one.

**Comment by sloriot** [comment]
Successfully tested in CGAL-6.2-Ic-108 (except on intel ICPC)

**Comment by MaelRL** [comment]
@lrineau please update CHANGES.md

**Comment by lrineau** [comment]
CHANGES updates by direct commit in `main`: https://github.com/CGAL/cgal/commit/7cf2d8ab369307cac642ce61a0da4efecb675479
